python/variables.py

```diff
@@ -0,0 +1,108 @@
+"""
+Copyright 2020 The Sensible Code Company Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from argparse import ArgumentParser
+import requests
+
+GRAPHQL_QUERY = """
+{
+ dataset(name: "%s"){
+  variables {
+  edges {
+   node {
+    name
+   }
+  }
+ }
+ ruleBase {
+  name
+  isSourceOf {
+   edges {
+     node {
+      name
+     }
+    }
+   }
+  }
+ }
+}
+"""
+
+
+def main():
+    """
+    Perform a GraphQL query to identify dataset variables.
+
+    Query the Cantabular extended API, verify and parse the response
+    and print the names of variables in the dataset.
+    """
+    args = parse_arguments()
+
+    query = GRAPHQL_QUERY % (args.dataset)
+    http_resp = requests.post(args.base_url, data={"query": query})
+    http_resp.raise_for_status()
+
+    resp = http_resp.json()
+    if "errors" in resp:
+        raise RuntimeError(resp["errors"][0]["message"])
+
+    # In Cantabular "rule variables" are special variables used for rule based redaction of
+    # categories when a cross tabulation is generated. Every rule variable is derived from
+    # the same base variable. Not all datasets have rule based redaction of categories. However,
+    # rule variables are still handled differently (e.g. by the UI) and a query may contain at
+    # most one rule variable which must be the first variable specified in a cross tabulation
+    # query.
+    rule_variables = {v["node"]["name"] for v in
+                      resp["data"]["dataset"]["ruleBase"]["isSourceOf"]["edges"]}
+    print("RULE VARIABLES")
+    print("--------------")
+    for variable in rule_variables:
+        print(variable)
+
+    variables = {v["node"]["name"] for v in resp["data"]["dataset"]["variables"]["edges"]}
+    print("")
+    print("OTHER VARIABLES")
+    print("---------------")
+    for variable in variables - rule_variables:
```

Comment:
I thought `variables - rule_variables` was a bit non-obvious. Maybe we should use brackets or assign to a separate variable?, e.g.
```
non_rule_variables = variables - rule_variables
```

Reply:
OK